test/framework/src/main/java/org/elasticsearch/transport/nio/channel/NioSocketChannel.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.transport.nio.ESSelector;
+import org.elasticsearch.transport.nio.SocketSelector;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SocketChannel;
+
+public class NioSocketChannel extends AbstractNioChannel<SocketChannel> {
+
+    private final InetSocketAddress remoteAddress;
+    private final ConnectFuture connectFuture = new ConnectFuture();
+    private WriteContext writeContext;
+    private ReadContext readContext;
+    private SocketSelector socketSelector;
+
+    public NioSocketChannel(String profile, SocketChannel socketChannel) throws IOException {
+        super(profile, socketChannel);
+        this.remoteAddress = (InetSocketAddress) socketChannel.getRemoteAddress();
+    }
+
+    @Override
+    public CloseFuture close(boolean attemptToCloseImmediately) {
+        // Even if the channel has already been closed we will clear any pending write operations just in case
+        if (attemptToCloseImmediately && state.get() > UNREGISTERED && selector.isOnCurrentThread() && writeContext.hasQueuedWriteOps()) {
+            writeContext.clearQueuedWriteOps(new ClosedChannelException());
+        }
+
+        return super.close(attemptToCloseImmediately);
+    }
+
+    @Override
+    public SocketSelector getSelector() {
+        return socketSelector;
+    }
+
+    @Override
+    public boolean markRegistered(ESSelector selector) {
+        this.socketSelector = (SocketSelector) selector;
+        return super.markRegistered(selector);
+    }
+
+    public int write(ByteBuffer buffer) throws IOException {
+        return socketChannel.write(buffer);
+    }
+
+    public long vectorizedWrite(ByteBuffer[] buffers) throws IOException {
+        return socketChannel.write(buffers);
+    }
+
+    public int read(ByteBuffer buffer) throws IOException {
+        return socketChannel.read(buffer);
+    }
+    public WriteContext getWriteContext() {
+        return writeContext;
+    }
+
+    public void setReadContext(TcpReadContext readContext) {
```

Comment:
I'll change it to one setting. The contexts normally need to take a reference to the channel. So the channel must first be constructed to created the contexts.